base/float.jl

```diff
@@ -165,36 +165,44 @@ promote_rule(::Type{Float16}, ::Type{UInt128}) = Float16
 promote_rule(::Type{Float16}, ::Type{Int128}) = Float16
 
 function Float64(x::UInt128)
-    A = 4.503599627370496e15 # Float64(UInt128(1) << 52)
-    B = 2.028240960365167e31 # Float64(UInt128(1) << 104)
-    C = 7.555786372591432e22 # Float64(UInt128(1) << 76)
-    D = 3.402823669209385e38 # Float64(typemax(UInt128))
-    if x < UInt128(1) << 104 # Can fit all info in two 52 bit mantissas
-        l = reinterpret(Float64, reinterpret(UInt64, A) | (x % UInt64) & significand_mask(Float64)) - A
-        h = reinterpret(Float64, reinterpret(UInt64, B) | ((x >> 52) % UInt64)) - B
-        l + h
-    else # Squish lowest bits to fit all info into two 52 bit mantissas, possible since lowest bits only affect rounding
-        l = reinterpret(Float64, reinterpret(UInt64, C) | ((x >> 24) % UInt64) & significand_mask(Float64) | (x % UInt64) & 0xffffff) - C
-        h = reinterpret(Float64, reinterpret(UInt64, D) | ((x >> 76) % UInt64)) - D
-        l + h
+    if x < UInt128(1) << 104 # Can fit in two 52 bits mantissa
+        low_exp = 4.503599627370496e15 # 2.0^52
+        high_exp = 2.028240960365167e31 # 2.0^104
```

Comment:
`0x1p104`

Reply:
Never seen this syntax, but could be nice.

Was a bit dissappointed when I couldn't write 2.0^52 since it seems the compiler does not take care of the calculation in most recent versions of julia, but in 1.7.3 it seems to do so. Not sure if this is something to report as a regression or if it is known and there for a reason?

1.7.3
```julia-repl
julia> f(x) = 2.0^52+x
f (generic function with 1 method)

julia> @code_llvm f(2.0)
;  @ REPL[1]:1 within `f`
define double @julia_f_392(double %0) #0 {
top:
; ┌ @ float.jl:399 within `+`
   %1 = fadd double %0, 0x4330000000000000
; └
  ret double %1
}
```
1.8.0-beta
```julia-repl
julia> f(x) = 2.0^52+x
f (generic function with 1 method)

julia> @code_llvm f(2.0)
;  @ REPL[1]:1 within `f`
define double @julia_f_637(double %0) #0 {
top:
; ┌ @ intfuncs.jl:326 within `literal_pow`
; │┌ @ math.jl:1037 within `^`
    %1 = call double @j_pow_body_639(double 2.000000e+00, i64 signext 52) #0
; └└
; ┌ @ float.jl:383 within `+`
   %2 = fadd double %1, %0
; └
  ret double %2
}
```